Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/TicketService.hs

```diff
@@ -594,3 +632,222 @@ postTicketBookingsUpdateSeats _ TicketBookingUpdateSeatsReq {..} = do
+cancelTicketBookingService ticketBooking tbServiceCancelReq = do
+  ticketBookingService <- QTBS.findByShortId tbServiceCancelReq.shortId >>= fromMaybeM (TicketBookingServiceNotFound tbServiceCancelReq.shortId.getShortId)
+  ticketService <- QTS.findById ticketBookingService.ticketServiceId >>= fromMaybeM (TicketServiceNotFound ticketBookingService.ticketServiceId.getId)
+  validateCancellationRequest ticketBookingService.status ticketService.allowCancellation
+  cancelledSeatsInServiceCategory <- processCancelServiceCategories ticketBooking.visitDate tbServiceCancelReq.serviceCategory
+  let (noOfSeatsToCancel, amountToRefund) = totalSeatsCancelledAndRefundAmount cancelledSeatsInServiceCategory
+
+  currentTime <- getCurrentTime
+  let cancellationPercentage = calculateCancellationCharges ticketBookingService ticketService currentTime ticketBooking.visitDate
+      refundAmount = case cancellationPercentage of
+        Just cancelCharge -> case cancelCharge.charge of
+          Domain.Types.TicketService.FlatFee fee -> amountToRefund - HighPrecMoney (fromIntegral fee)
+          Domain.Types.TicketService.Percentage per -> amountToRefund - ((HighPrecMoney (fromIntegral per) * amountToRefund) / 100)
+        Nothing -> amountToRefund
+
+  pure refundAmount
+  where
+    totalSeatsCancelledAndRefundAmount = foldl (\(seats, amount) (s, a) -> (seats + s, amount + a)) (0, HighPrecMoney 0)
+    validateCancellationRequest status allowCancellation = unless (status == DTB.Confirmed && allowCancellation) $ throwError $ InvalidRequest "Ticket Cancellation Not allowed"
+    processCancelServiceCategories visitDate = mapM (cancelTBServiceCategory visitDate)
+
+calculateCancellationCharges :: DTB.TicketBookingService -> Domain.Types.TicketService.TicketService -> UTCTime -> Data.Time.Calendar.Day -> Maybe Domain.Types.TicketService.CancellationCharge
+calculateCancellationCharges ticketBookingService ticketService currentTime visitDate =
+  case ticketService.cancellationCharges of
+    Just (cancellationCharges' :: [Domain.Types.TicketService.CancellationCharge]) ->
+      let cancellationCharges = sortBy (\a1 a2 -> compare a1.time a2.time) cancellationCharges'
+          startTime = case ticketBookingService.btype of
+            Domain.Types.BusinessHour.Slot time -> time
+            Domain.Types.BusinessHour.Duration startTime' _ -> startTime'
+          visitDateTime = UTCTime visitDate (timeOfDayToTime startTime)
+          remainingSeconds = Kernel.Prelude.roundToIntegral $ diffUTCTime currentTime visitDateTime --(Data.Time.Calendar.diffDays ticketBookingService.visitDate (getCurrentDate currentTime) * 86400) + (fromIntegral $ timeOfDayToTime startTime)
+          idx = fromMaybe (length cancellationCharges) $ findIndex (\item -> item.time DO.> remainingSeconds) cancellationCharges
+          index = if idx == 0 then idx else idx - 1
+       in cancellationCharges !? index
```

Comment:
Just sort the `cancellationCharge` array w.r.t. decreasing order of time and use `find` function to get the first time > remainingSeconds. In config, we can have like `[(0, 24 hrs), (75, 12 hrs), (25, 4 hrs)]`. Let's say you're cancelling before 16 hrs, so `find` will give (75, 12 hrs)